src/engine/rendering/gui/GuiManager.cpp

```diff
@@ -0,0 +1,59 @@
+#include "GuiManager.hpp"
+
+#include "components/GuiComponent.hpp"
+#include "core/Application.hpp"
+
+namespace lei3d 
+{
+	GuiManager* GuiManager::s_instance = new GuiManager();
+
+	GuiManager& GuiManager::Instance()
+	{
+		return *s_instance;
+	}
```

Comment:
Just curious, what would make the original impl. unsafe? Is it bc s_instance can not be initialized? (I actually don't know when s_instance initializes since usually static vars are init at compile time, but can you actually initialize something at compile time using new? I'm guessing it's not lazy init.).

Reply:
Not gonna say I know for sure with this implementation but to be thread-safe, you'll need locks. The original implementation is doing something like [double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking). The change I made is the preferred way to do C++ singleton patterns since C++11, which is supported by modern compilers. The first time program flow enters the function the singleton instance is created and other concurrent flows will wait for it.

In this case, the memory is allocated at run-time rather than compile time and it also doesn't invoke the destructor on program termination because it was created  with `new`.